### ISSUE TITLE
Fix S4143: False positive on dictionaries identified by multiple member access expressions

### ISSUE THREAD

**Issue by valhristov** [SonarSource/sonar-dotnet#1897]
### Description

The rule raises on the last two lines in the following code snippet:

```
CloudDrizzleImage.ReplaceStringMap.Add("currentColor", "#FFFF0000");
CloudLightningImage.ReplaceStringMap.Add("currentColor", "#FF00FF00"); // FP
CloudRainImage.ReplaceStringMap.Add("currentColor", "#FF0000FF"); // FP
```

This is because when the dictionary is identified by two-levels of member access expressions we take into account only the identifier name of the second one. In other words, we treat the code above as if it was:

```
ReplaceStringMap.Add("currentColor", "#FFFF0000");
ReplaceStringMap.Add("currentColor", "#FF00FF00");
ReplaceStringMap.Add("currentColor", "#FF0000FF");
```

### Related information

* SonarC# Version 7.3
* Reported in SUPPORT-11285

**Comment by Evangelink** [comment]
Hi @valhristov,

I cannot reproduce the issue so I am closing this ticket.

**Comment by nk111** [comment]
I see the same issue when analyzing a vb.net asp.net project on SQ 7.4.0.18908

Basically 1/3 of all found bugs are false positive S4143...

We are using the SonarQube Azure DevOps Marketplace Extension to integrate SQ into our on-premise TFS 2018. Any hints what to do?

**Comment by Evangelink** [comment]
Hi @nk111,

Could you let us know the version of SonarVB you are using? Also would you mind creating a small reproducer on which an issue would be raised so that we can have a look at it?

Cheers,
Amaury

**Comment by nk111** [comment]
Sure @Evangelink ,

i just don't know exactly where to find the correct version information for the VB scanner? Here are some lines from the "Run code analysis" task log on TFS:

```
2019-01-09T22:29:11.7655115Z Description  : Run scanner and upload the results to the SonarQube server.
2019-01-09T22:29:11.7655115Z Version      : 4.4.1
2019-01-09T22:29:13.2655115Z SonarScanner for MSBuild 4.4.2
2019-01-09T22:29:13.2811365Z Using the .NET Framework version of the Scanner for MSBuild
2019-01-09T22:29:20.0096515Z INFO: SonarQube Scanner 3.2.0.1227
2019-01-09T22:29:20.7284015Z INFO: SonarQube server 7.4.0
2019-01-09T22:30:56.4002765Z INFO: Sensor VB.NET Properties [vbnet]
```

If you need any other version information please let me know where to look.


Here is an example of a false positive S4134 in a .vb file:

```
btn_ChangeNative.Attributes.Item("Title") = "some text goes here"
btn_ChangeForeign.Attributes.Item("Title") = "some other text goes here"
```

and another one:

```
Me.btn_CancelChangedData.Style.Item("display") = "none"
Me.btn_AcceptChangedData.Style.Item("display") = "none"
```

In both cases we get a "Verify this is the index/key that was intended; a value has already been set for it."

**Comment by Evangelink** [comment]
You could try to go to the following URL `<YOUR_SQ_URL>/api/plugins/installed` and look for `"name": "SonarVB"` which should have an entry with the version.

Could you let me know if the code sample referenced in #2235 does raise an issue for you? Because for now, I am having some trouble to reproduce the issue.

**Comment by nk111** [comment]
Here is the version information:

`{"key":"vbnet","name":"SonarVB","description":"Code Analyzer for VB.NET","version":"7.7 (build 7192)","license":"GNU LGPL 3","organizationName":"SonarSource","organizationUrl":"http://www.sonarsource.com","editionBundled":false,"homepageUrl":"http://redirect.sonarsource.com/plugins/vbnet.html","issueTrackerUrl":"https://github.com/SonarSource/sonar-csharp/issues","implementationBuild":"c46d44cdae6e75538626f380fe3fa7553e6cf93d","updatedAt":1541579526343,"filename":"sonar-vbnet-plugin-7.7.0.7192.jar","sonarLintSupported":false,"hash":"689ca1ee3585448284954dd0e15d0198"}`

And i was able to build your test code with some modification:

```
Imports System.Collections.Generic

Module SonarTestModule

   Sub Main()
      Dim f1 As New Foo
      Dim f2 As New Foo

      f1.Attributes.Item("Title") = "some text goes here"
      f2.Attributes.Item("Title") = "some other text goes here"
   End Sub

End Module

Class Foo
   Public Attributes As Dictionary(Of String, String)
End Class
```
And here is the result:

![sonar](https://user-images.githubusercontent.com/155322/51121206-28d6f000-1817-11e9-93f3-401c9eaa687e.PNG)

**Comment by Evangelink** [comment]
Hi @nk111,

So I can confirm that in 7.10 the bug you are reporting doesn't exist and was correctly fixed. We do have another problem (linked to the other ticket) that causes the analyzer to crash.

I would suggest you to follow the resolution of the other ticket and then to update (or ask your admin to update) the version of SonarVB you are using.

**Comment by nk111** [comment]
Which other ticket are you talking about? #2235 ? 

And what exactly would i need to update? I guess its server side. So should an update from 7.4.0.18908 to 7.5 already fix our issue? Or is it possible to update the VB plugin only?

**Comment by Evangelink** [comment]
As you can see with the milestone, the FP fix was done in version 7.10 so you will need to update SonarVB to the version 7.10. Sadly if you are doing so, you will end up with some missing issues because of the bug described in #2235.

If you update right now to SonarVB 7.10 you will be able to get rid of all the FPs you are mentioning. Note that we might miss a few wrong places because we might end up in situations described in #2235 causing the analyzer not to report an issue.

Is it clear?

**Comment by nk111** [comment]
Thanks for the info. Just one question left. How exactly do i update a single plugin like SonarVB to a higher version?

**Comment by nk111** [comment]
Nevermind! Just found the update option in the server backend ;) Sorry, this is all new to me!

**Comment by Evangelink** [comment]
No worries ;)
